iup/all_objects.py

```diff
@@ -0,0 +1,366 @@
+
+        assert all(
+            isinstance(rollout_date, dt.date) for rollout_date in self.rollout_dates
+        ), "Rollout dates contain a non-date item."
+
+        assert isinstance(
+            self.test_data, CumulativeUptakeData
+        ), "Test data are not cumulative uptake data."
+        assert isinstance(
+            self.train_data, IncidentUptakeData
+        ), "Training data are not incident uptake data."
+
+    def build_model(self) -> Self:
+        this_model = LinearIncidentUptakeModel()
+        this_model = this_model.fit(self.train_data)
+        this_model = this_model.predict(
+            self.train_data,
+            self.start_date,
+            self.end_date,
+            self.interval,
+            self.init_elapsed,
+            self.init_interval,
+            self.init_incident,
+            self.init_cumulative,
+        )
+
+        setattr(self, "model", this_model)
+        return self
+
+
+class UptakeModel(abc.ABC):
+    @abc.abstractmethod
+    def fit(self, data: UptakeData) -> Self:
+        pass
+
+    @abc.abstractmethod
+    def predict(self, data: UptakeData, *args, **kwargs) -> UptakeData:
+        pass
+
+
+class LinearIncidentUptakeModel(UptakeModel):
+    def __init__(self):
+        self.model = LinearRegression()
+        pass
+
+    def fit(self, train: IncidentUptakeData) -> Self:
+        assert isinstance(train, IncidentUptakeData)
+        train.validate()
+
+        train_std = train.with_columns(
+            previous_std=(pl.col("previous") - pl.mean("previous"))
+            / pl.std("previous"),
+            elapsed_std=(pl.col("elapsed") - pl.mean("elapsed")) / pl.std("elapsed"),
+            estimate_std=(pl.col("estimate") - pl.mean("estimate"))
+            / pl.std("estimate"),
+        )
+
+        x = (
+            train_std.select(["previous_std", "elapsed_std"])
+            .with_columns(interact=pl.col("previous_std") * pl.col("elapsed_std"))
+            .to_numpy()
+        )
+        y = train_std.select(["estimate_std"]).to_numpy()
+
+        self.model.fit(x, y)
+
+        return self
+
+    def predict(
+        self,
+        train: IncidentUptakeData,
+        start_date,
+        end_date,
+        interval,
+        init_elapsed,
+        init_interval,
+        init_incident,
+        init_cumulative,
+    ) -> Self:
+        previous_mn = train["previous"].mean()
+        previous_sd = train["previous"].std()
+        elapsed_mn = train["elapsed"].mean()
+        elapsed_sd = train["elapsed"].std()
+        estimate_mn = train["estimate"].mean()
+        estimate_sd = train["estimate"].std()
+
+        incident_proj = (
+            pl.date_range(
+                start=start_date,
+                end=end_date,
+                interval=interval,
+                eager=True,
+            )
+            .alias("date")
+            .to_frame()
+            .with_columns(
+                region=pl.lit(train.select(pl.col("region").unique())),
```

Comment:
Again, I disagree. While this is currently set up to fit a model on only one curve from one region and one season, a _much_ better statistical model will be a multilevel one fit across multiple seasons (and maybe multiple regions) with group-level hyperparameters for each. I want the groundwork for this already laid (mistake of assuming just 1 region notwithstanding) from the beginning.

Reply:
great opportunity to clarify! The *current* model has no hierarchy/pooling, and so I claim it should not be looking for things like region. We should be able to pass it a data frame that doesn't have region in it and have it still work.

Agree that we'll want other models that *require* region (or some other kind of grouping) in the input data. We'll want yet other ones that have year (but not region). And then some that have season *and* region. And then, further down the line, we'll also want ones that have population subgroups!

In the variant nowcasting work, we called this a "ladder" of models, from simpler to more complex.

All the model objects should have some things about their schema the same, but not necessarily exactly the same table schema.

In the variant nowcasting work, we had one model that took only one state's data, and then another one that took multiple states' data and treated them all as independent. Those two approaches should yield the same results, but when they didn't it was a useful way to debug a numpyro problem

Reply:
There's a third option, besides no-pooling and hierarchical-pooling, which is complete-pooling. In complete-pooling, there may be multiple regions, but all regions are assumed to behave identically (i.e. uptake in all regions is governed by the same non-hierarchical regression parameters). For example, if your data set includes 2 different states, a complete-pooling model would essentially ignore state to learn from twice as much data to infer 1 set of parameters. Then it would use that 1 set of parameters to make projections for the 2 states separately. (The only way the projections for the 2 states could differ is if they have different initial uptake.)

In this case, no-pooling is a special case of complete-pooling. The same exact code that builds a complete-pooling model will naturally also build a no-pooling model if there is only one region represented in the data set. So we don't need to build extra models that explicitly expect region-less data. Instead, we just build a complete-pooling model expects data with region, and it will work seamlessly whether there is just one or more than one region.